### ISSUE TITLE
[Question] Front direction fault of SwitchT and unused value of Switch

### ISSUE THREAD

**Issue by JunmoreHeavyBox** [leezer3/OpenBVE#997]
### Description

After the train passes forward through SwitchT, the train does not reach the first track, but runs on the void track extended by the last section of the breakpoint track, and the frame rate drops.

I'm going to name a switch, I need to write it this way.
Track.Switch 0; 3; 0; _~Unused~_; Crossover Down; Main; To Down
Track.SwitchT 0; 2; 0; _~Unused~_; Crossover Up; Main; To Up
I'm not sure if the unused value was intentionally reserved or an error?

### Related information 

* OpenBVE Version: Build 2024-02-17

**Comment by leezer3** [comment]
That was intended to control spring return switches.
Haven't implemented it yet as it got a little complex. to make work correctly.

Must update the documentation, thanks....

**Comment by JunmoreHeavyBox** [comment]
![Looks like derailment 1](https://github.com/leezer3/OpenBVE/assets/114210303/ea1c2d78-9425-40e5-82d0-b82992df679b)
![Looks like derailment 2](https://github.com/leezer3/OpenBVE/assets/114210303/806c9430-0f89-4476-ae05-4283cdecbc3e)

Interesting, just like the derailment.

The SwitchT and tracks were placed according to the docs instructions and the train looked correct after driving forward. I wanted to make compact switches and didn't want to extend the tracks. So I undid the extension and placed the SwitchT in the same position as the RailEnd. Then the train entered the vold track after passing forward.

- As I drove backwards through the SwitchT in the vold track.
  - SwitchT's position is after or equal to RailEnd, the train can only go to the second track, the source track of the vold track.
  - SwitchT's position precedes RailEnd for the switch to work properly. This is of course for trains travelling on the vold track.

- In any case, the train travelling forward past SwitchT goes to the vold track until it encounters the same RailStart at the index of the vold track again.

- I reversed my train from the other track back to the main track and drove backwards past SwitchT as if SwitchT didn't exist. And I heard the Switch Sound twice.
  - At least Switch is normal, enabling me to do this.

- I think it's the strange SwitchT implementation method, the one through the vold track.

It seems that SwitchT has the first and second track names reversed?

I want to add the track index where the player train is located in the debug interface.


### Reproduction

## Route

Here is my route files.
[Switch Test Routes.zip](https://github.com/leezer3/OpenBVE/files/14464802/Switch.Test.Routes.zip)

**Comment by JunmoreHeavyBox** [comment]
Please have a check at this.

**Comment by JunmoreHeavyBox** [comment]
After passing through SwitchT, it will still goes to the void track.
Require adding a track index in the debug interface.

**Comment by JunmoreHeavyBox** [comment]
@leezer3 Add track index in the debugging interface. I said it three times.

**Comment by leezer3** [comment]
Again, there is only one of me, and I've got limited free time to work on this......

I haven't got as far as investigating the void track issue properly yet, but will try and do so.

**Comment by leezer3** [comment]
Made a few changes for the current build.

I think this probably fixes what you were seeing (running through a reverse switch with the 'wrong' direction set could get into an invalid element).

**Comment by JunmoreHeavyBox** [comment]
The train still went to the void track:<

But I will open some large issues...

**Comment by JunmoreHeavyBox** [comment]
The recent build has not resolved, and the train located on the second track still went to the void track.

**Comment by leezer3** [comment]
Current build makes most things OK.

However, the use of the .Turn command and the .Switch / .SwitchT command in a routefile are no longer supported, and an error will now be generated to that effect.
TLDR:
The .Turn command rotates the entire world, and messes with track lengths (by inserting additional elements).
Please use the .Curve command instead.

**Comment by JunmoreHeavyBox** [comment]
Got it.

**Comment by JunmoreHeavyBox** [comment]
I tried moving RailEnd forward one block before I found out about it.
The second track issue only fixes the last SwitchT in the same second track index.

**Comment by JunmoreHeavyBox** [comment]
It seems that except for the SwitchT on rail 0, the initial setting of SwitchT on other tracks is opposite.

**Comment by leezer3** [comment]
I wouldn't reccommend re-using rail indicies at all where switches are concerned- Keep anything drivable completely separate.

Track elements are a linear array, which really doesn't work too well for stopping / starting them.
It's on the list to re-do this (essentially, a track element should only link to the previous / next).

Looking into the SwitchT reversal too.

**Comment by JunmoreHeavyBox** [comment]
I don't like it, reusing the track index is convenient for me to layout.

I don't care about how the track starts and ends. For switch issues, in my opinion, the switch behavior is roughly like this.
driving in the direction of convergence, it is similar to passing through `.PlayerPath ST;FT, .PlayerPath FT;FT`, change the track when driving in the direction of separation.

I am also considering using PlayerPath to solve the problem. I have made a pseudo spring switch before.

**Comment by JunmoreHeavyBox** [comment]
Just about enough.
